### PR TITLE
chore(clients): Replace existing Signer implementation

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR replaces the underlying implementation of the existing Signer with new SigV4 signing functions without changing
1. The Signer's overall class structure
2. The Signer's static function signatures
3. The Signer's static function output

#### Description of how you validated changes
1. Adapted Signer unit tests to more comprehensively test various input/output expectations
2. Swapped out Signer implementation
3. Made sure the same unit test suites passed
4. Tested the some call sites to validate signing was happening and services were not rejecting the requests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-rest/__tests__/RestAPI-test.ts [hunk 1]
```diff
@@ -514,7 +514,7 @@ describe('Rest API test', () => {
 				endpoints: [
 					{
 						name: 'url',
-						endpoint: 'endpoint',
+						endpoint: 'https://domain.fakeurl/',
```
Comment (cshfang) [inline packages/api-rest/__tests__/RestAPI-test.ts:517]: Signer requires parse-able URLs

File: packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts [hunk 2]
```diff
@@ -430,7 +430,7 @@ describe('AnalyticsProvider test', () => {
 				await analytics.record(params, { resolve, reject });
 
 				const expectedUrl =
-					'https://pinpoint.region.amazonaws.com/v1/apps/appId/events/legacy?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=accessKeyId%2FisoStrin%2Fregion%2Fmobiletargeting%2Faws4_request&X-Amz-Date=isoString&X-Amz-Security-Token=sessionToken&X-Amz-SignedHeaders=host&X-Amz-Signature=9dfa2a29782d344c56a9ab99fe58db6d1748e097ae418c398b26ab372a23f22f';
+					'https://pinpoint.region.amazonaws.com/v1/apps/appId/events/legacy?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=accessKeyId%2FisoStrin%2Fregion%2Fmobiletargeting%2Faws4_request&X-Amz-Date=isoString&X-Amz-SignedHeaders=host&X-Amz-Security-Token=sessionToken&X-Amz-Signature=9dfa2a29782d344c56a9ab99fe58db6d1748e097ae418c398b26ab372a23f22f';
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts:433]: No these are appended after calculating the signature

**Review by AllanZhengYP — APPROVED** [review]
The change looks good to me. The bundle size has minor increase for other packages. The root cause is unknown still, but all the bundle sizes still have significant net decrease compared to the main branch.

**Review by jimblanc — APPROVED** [review]
Looks good

**Pull request merged**